File: web-service/src/main/java/obby0511/sentiment/analysis/Handler.java
package obby0511.sentiment.analysis;

import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.server.ServerRequest;
import org.springframework.web.reactive.function.server.ServerResponse;
import reactor.core.publisher.Mono;

import java.nio.charset.StandardCharsets;

import static org.springframework.web.reactive.function.BodyInserters.fromPublisher;
import static org.springframework.web.reactive.function.server.ServerResponse.notFound;
import static org.springframework.web.reactive.function.server.ServerResponse.ok;

@Component
public class Handler {
    private final LanguageService language = new LanguageServiceImpl();

    public Mono<ServerResponse> handle(ServerRequest request) {
        return ok().body(fromPublisher(request.formData().map(data ->
                language.analyzeSentiment(LanguageService.LanguageRequest.builder()
                        .text(data.getFirst("text"))
                        .encodingType(StandardCharsets.UTF_8)
                        .build()
                )), LanguageService.LanguageResponse.class))
                .switchIfEmpty(notFound().build());
    }
}

File: web-service/src/main/kotlin/obby0511/sentiment/analysis/Handler.kt
package obby0511.sentiment.analysis

import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.BodyInserters.fromPublisher
import org.springframework.web.reactive.function.server.ServerRequest
import org.springframework.web.reactive.function.server.ServerResponse
import reactor.core.publisher.Mono

@Component
class Handler {

    val language: LanguageService = DefaultLanguageService()

    fun analyze(request: ServerRequest): Mono<ServerResponse> {
        val scores = fromPublisher(request
                .bodyToMono(LanguageRequest::class.java)
                .map { text -> language.analyzeSentiment(text) },
                LanguageResponse::class.java)
        return ServerResponse.ok().body(scores)
    }
}

